File: src/cadenza/application/resolve.py
"""Name -> ResolvedProject (design section 6)."""

from __future__ import annotations

import difflib

from cadenza.application.compose import Catalog
from cadenza.domain.digest import config_digest
from cadenza.domain.errors import ProjectNotFoundError
from cadenza.domain.project import ResolvedProject

_MAX_SUGGESTIONS = 5


def resolve_project(catalog: Catalog, name: str) -> ResolvedProject:
    project_id = catalog.names.get(name)
    if project_id is None:
        raise ProjectNotFoundError(_not_found_message(catalog, name))
    project = catalog.projects[project_id]
    return ResolvedProject(
        project_id=project.project_id,
        aliases=project.aliases,
        source=project.source,
        base_branch=project.base_branch,
        # The digest is what lets a later reader tell that the catalog moved
        # under a run that already happened (design section 3.2).
        config_digest=config_digest(project),
        provenance=catalog.provenance[project_id],
    )


def _not_found_message(catalog: Catalog, name: str) -> str:
    known = sorted(catalog.names)
    suggestions = difflib.get_close_matches(name, known, n=_MAX_SUGGESTIONS)
    if not suggestions:
        return f"no project is named '{name}'"
    return f"no project is named '{name}'. Closest known names: {', '.join(suggestions)}"

File: src/application/resolve.ts
/**
 * Name -> ResolvedProject (design doc section 6).
 */
import { compareByCodePoint } from "../domain/canonical-json.js";
import { configDigest } from "../domain/digest.js";
import { ProjectNotFoundError } from "../domain/errors.js";
import type { ResolvedProject } from "../domain/project.js";
import { getCloseMatches } from "../domain/python-difflib.js";
import type { Catalog } from "./compose.js";

const MAX_SUGGESTIONS = 5;

export function resolveProject(catalog: Catalog, name: string): ResolvedProject {
  const projectId = Object.hasOwn(catalog.names, name) ? catalog.names[name] : undefined;
  if (projectId === undefined) {
    throw new ProjectNotFoundError(notFoundMessage(catalog, name));
  }
  const found = catalog.projects[projectId] as (typeof catalog.projects)[string];
  return Object.freeze({
    projectId: found.projectId,
    aliases: found.aliases,
    source: found.source,
    baseBranch: found.baseBranch,
    // The digest is what lets a later reader tell that the catalog moved under a
    // run that already happened (design doc section 3.2).
    configDigest: configDigest(found),
    provenance: catalog.provenance[projectId] as Readonly<
      Record<string, (typeof catalog.provenance)[string][string]>
    >,
  });
}

function notFoundMessage(catalog: Catalog, name: string): string {
  // Python's `sorted()` on a dict yields its keys in code-point order, which is
  // also the order the suggestions are drawn from.
  const known = Object.keys(catalog.names).sort(compareByCodePoint);
  const suggestions = getCloseMatches(name, known, MAX_SUGGESTIONS);
  if (suggestions.length === 0) {
    return `no project is named '${name}'`;
  }
  return `no project is named '${name}'. Closest known names: ${suggestions.join(", ")}`;
}
